musicgen/musicgen.py

```diff
@@ -0,0 +1,359 @@
+        for offset in tqdm(range(max_steps)):
+            audio_input = mx.tile(audio_seq[:, offset : offset + 1], [2, 1, 1])
+            audio_logits = self(audio_input, text_tokens, cache)
+            cond_logits, uncond_logits = audio_logits[:1], audio_logits[1:2]
+            audio_logits = uncond_logits + (cond_logits - uncond_logits) * guidance_coef
+            audio_tokens = top_k_sampling(audio_logits, top_k, temp, axis=-2)
+            # "delay" pattern
+            audio_tokens[..., offset + 1 :] = self.bos_token_id
+            audio_tokens[..., : -max_steps + offset] = self.bos_token_id
+            audio_seq[:, offset + 1 : offset + 2] = audio_tokens
+            mx.eval(audio_seq)
+
+        # Undo delay
+        for i in range(self.num_codebooks):
+            audio_seq[:, : -self.num_codebooks, i] = audio_seq[
+                :, i : -self.num_codebooks + i, i
+            ]
+        audio_seq = audio_seq[:, 1 : -self.num_codebooks + 1]
+
+        audio_seq = mx.swapaxes(audio_seq, -1, -2)[:, mx.newaxis]
+        audio = self._audio_decoder.decode(audio_seq, audio_scales=[None])
+        return audio[0]
+
+    @classmethod
+    def sanitize(cls, weights):
+        out_weights = {}
+        for k, arr in weights.items():
+            if k.startswith("transformer."):
+                k = k[len("transformer.") :]
+
+            if "cross_attention" in k:
+                k = k.replace("cross_attention", "cross_attn")
+
+            if "condition_provider" in k:
+                k = k.replace(
+                    "condition_provider.conditioners.description", "text_conditioner"
+                )
+
+            if "in_proj_weight" in k:
+                dim = arr.shape[0] // 3
+                name = "in_proj_weight"
+                out_weights[k.replace(name, "q_proj.weight")] = arr[:dim]
+                out_weights[k.replace(name, "k_proj.weight")] = arr[dim : dim * 2]
+                out_weights[k.replace(name, "v_proj.weight")] = arr[dim * 2 :]
+                continue
+
+            out_weights[k] = arr
+        return out_weights
+
+    @classmethod
+    def from_pretrained(cls, path_or_repo: str):
+        import torch
+        from huggingface_hub import snapshot_download
+
+        path = Path(path_or_repo)
+        if not path.exists():
+            path = Path(
+                snapshot_download(
+                    repo_id=path_or_repo,
+                    allow_patterns=["*.json", "state_dict.bin"],
+                )
+            )
+
+        with open(path / "config.json", "r") as f:
+            config = SimpleNamespace(**json.load(f))
+            config.text_encoder = SimpleNamespace(**config.text_encoder)
+            config.audio_encoder = SimpleNamespace(**config.audio_encoder)
+            config.decoder = SimpleNamespace(**config.decoder)
+
+        weights = torch.load(path / "state_dict.bin", weights_only=True)["best_state"]
+        weights = {k: mx.array(v.numpy()) for k, v in weights.items()}
```

Comment:
I think it's better to go directly to `mx.array` here and not step through `numpy`. Avoids a copy and also won't crash for `bfloat16` (not sure if that's a thing)

Reply:
Really nice that that works!